### PR TITLE
EZP-27428: Implemented MultiLanguage logic for ObjectStateService

### PR DESCRIPTION
**Pull request by alongosz**
Part of [EZP-27428](https://jira.ez.no/browse/EZP-27428) implementation for #2002
----
**Target branch**: `language_match_api_logic`

This PR refactors ObjectState-related code to implement multi-language / unified logic introduced in #2002.

**TODO**:
- [x] Implement unified language for `ObjectStateService`.
- [x] Refactor `ObjectStateService` integration tests to use unified names and descriptions (so test code can be reused for new tests).
- [x] Add `ObjectStateService` integration tests for new multi-language / unified logic.
- [x] Align other existing tests with the changes.
- [x] Optimize Core `ObjectStateService::getContentState`.
- [x] Remove TMP commit.

### CODE REVIEW

File: eZ/Publish/API/Repository/Tests/ObjectStateServiceTest.php [hunk 1]
```diff
@@ -229,11 +229,11 @@ public function testCreateObjectStateGroup()
         $objectStateGroupCreate->defaultLanguageCode = 'eng-US';
         $objectStateGroupCreate->names = array(
             'eng-US' => 'Publishing',
-            'eng-GB' => 'Sindelfingen',
+            'ger-DE' => 'Sindelfingen',
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/ObjectStateServiceTest.php]: Note: this is just for unification of names and descriptions to use common data provider for multi-language / unified logic.

File: eZ/Publish/API/Repository/Tests/ObjectStateServiceTest.php [hunk 2]
```diff
@@ -1017,6 +1022,188 @@ public function testLoadObjectStateThrowsNotFoundException()
...
+     * @dataProvider getPrioritizedLanguagesList
+     * @param string[] $prioritizedLanguages
+     * @param string|null $expectedLanguageCode
+     */
+    public function testLoadObjectStateGroupsWithPrioritizedLanguagesList(
+        array $prioritizedLanguages,
+        $expectedLanguageCode
+    ) {
+        // cleanup before the actual test
+        $this->deleteExistingObjectStateGroups();
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/ObjectStateServiceTest.php]: Note: already existing in initial data object state group messes with this test, so I removed it (just for this single test).
As an alternative, I could've improved initial data, but I found more valuable to test if what API just created aligns with what gets loaded.

File: eZ/Publish/Core/REST/Client/ObjectStateService.php [hunk 3]
```diff
@@ -251,15 +233,9 @@ public function createObjectState(ObjectStateGroup $objectStateGroup, ObjectStat
...
-     *
-     * @param mixed $stateId
-     *
-     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException if the state was not found
-     *
-     * @return \eZ\Publish\API\Repository\Values\ObjectState\ObjectState
+     * {@inheritdoc}
      */
-    public function loadObjectState($stateId)
+    public function loadObjectState($stateId, array $prioritizedLanguages = [])
```
Comment (alongosz) [inline eZ/Publish/Core/REST/Client/ObjectStateService.php]: @andrerom: I'm not sure how to pass prioritized languages to REST request. However you've mentioned that there are parts related to REST Client that we don't support (are broken), right? So maybe I shouldn't worry?

File: eZ/Publish/Core/Repository/ObjectStateService.php [hunk 4]
```diff
@@ -499,7 +492,7 @@ public function getContentState(ContentInfo $contentInfo, APIObjectStateGroup $o
             $objectStateGroup->id
         );
 
-        return $this->buildDomainObjectStateObject($spiObjectState);
+        return $this->buildDomainObjectStateObject($spiObjectState, $objectStateGroup);
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/ObjectStateService.php]: Note: this is not directly related to changes, just some optimization. If 2nd argument to `buildDomainObjectStateObject` is `null`, then `ObjectStateGroup` is loaded from the persistence layer, which is redundant.

File: eZ/Publish/Core/Repository/ObjectStateService.php [hunk 5]
```diff
@@ -588,28 +585,33 @@ protected function buildDomainObjectStateObject(SPIObjectState $spiObjectState,
                 'names' => $spiObjectState->name,
                 'descriptions' => $spiObjectState->description,
                 'objectStateGroup' => $objectStateGroup,
-            )
+                'prioritizedLanguages' => $prioritizedLanguages,
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/ObjectStateService.php]: Note: `ObjectState` (and `ObjectStateGroup`) uses different naming for some properties. 
There exists `languageCodes` property, however I decided to add `prioritizedLanguages` separately as the `languageCodes` comes from persistence layer and maybe should be left alone. // cc @andrerom 

On the other hand we already have `defaultLanguage` property, so I didn't add extra one `mainLanguageCode`.

File: eZ/Publish/Core/Repository/Values/ObjectState/ObjectState.php [hunk 6]
```diff
@@ -26,16 +23,33 @@
  */
 class ObjectState extends APIObjectState
 {
-    use MultiLanguageTrait;
-    use MultiLanguageNameTrait;
-    use MultiLanguageDescriptionTrait;
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Values/ObjectState/ObjectState.php]: Note: Actually we can't use these traits here because, as mentioned above, `mainLanguageCode` is called here `defaultLanguageCode`

File: eZ/Publish/Core/Repository/Values/ObjectState/ObjectStateGroup.php [hunk 7]
```diff
@@ -25,7 +22,76 @@
  */
 class ObjectStateGroup extends APIObjectStateGroup
 {
-    use MultiLanguageTrait;
-    use MultiLanguageNameTrait;
-    use MultiLanguageDescriptionTrait;
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Values/ObjectState/ObjectStateGroup.php]: Note: same comment as above

File: eZ/Publish/Core/SignalSlot/Tests/ObjectStateServiceTest.php [hunk 8]
```diff
@@ -55,145 +55,145 @@ public function serviceProvider()
...
-            array(
+            ],
+            [
                 'newObjectStateCreateStruct',
-                array('identifier'),
+                ['identifier'],
                 $objectStateCreateStruct,
                 0,
-            ),
-        );
```
Comment (alongosz) [inline eZ/Publish/Core/SignalSlot/Tests/ObjectStateServiceTest.php]: Note: sorry for this big block of short-array-syntax changes. I aligned this test with extra empty array parameters and it made sense to change the whole block rather than mix short with long array syntax.

File: eZ/Publish/Core/REST/Client/ObjectStateService.php (hunk 3 above)
Comment (andrerom) [inline eZ/Publish/Core/REST/Client/ObjectStateService.php]: Don't worry, be happy :)

File: eZ/Publish/Core/Repository/Values/ObjectState/ObjectState.php (hunk 6 above)
Comment (andrerom) [inline eZ/Publish/Core/Repository/Values/ObjectState/ObjectState.php]: what about using a *(internal documented)* constant that can be set to the fallback language propert, _main_ (defualt), _inital_ (versionInfo, but not applicable here) and _default_ in this case

Comment (alongosz) [inline eZ/Publish/Core/Repository/Values/ObjectState/ObjectState.php]: > what about using a (internal documented) constant that can be set to the fallback language propert, main (defualt), inital (versionInfo, but not applicable here) and default in this case

Meaning e.g. 
`const DEFAULT_LANGUAGE_PROPERTY = 'defaultLanguageCode';`
with usage as:
`$this->names[$this->{static::DEFAULT_LANGUAGE_PROPERTY}]`
? :)

Could be, it's PHP after all ;-)
I've created some unit tests as this logic seems to get more complicated and if you're fine with it I'll include them in this PR.

Comment (gggeek) [inline eZ/Publish/Core/Repository/Values/ObjectState/ObjectState.php]: why `$this->{static::DEFAULT_LANGUAGE_PROPERTY}` instead of `static::DEFAULT_LANGUAGE_PROPERTY` ?

Comment (alongosz) [inline eZ/Publish/Core/Repository/Values/ObjectState/ObjectState.php]: @gggeek It holds the actual property name "resolved" dynamically, so in this case it would mean
`$this->defaultLanguageCode`
but in case of other ValueObjects, like `ContentType` it would be resolved as:
`$this->mainLanguageCode`

This confusion is the main reason I'm rather against making dynamically accessed properties. Moreover it confuses IDE and makes code maintenance more difficult.

However it's ValueObject which by definition uses dynamic properties and also it's internal, so in this case I'm fine with it as long as this is what @andrerom implied :)

Comment (gggeek) [inline eZ/Publish/Core/Repository/Values/ObjectState/ObjectState.php]: doh - I get it now (I thought that the 'dynamic part' referred to using `static` instead of `self`...).
Tbh I find this syntax highly confusing.

Comment (gggeek) [inline eZ/Publish/Core/Repository/Values/ObjectState/ObjectState.php]: What about instead deprecating property 'defaultLanguageCode', making it available via magic __get() for BC, and switch over to using 'mainLanguageCode' even for obj states and stategroups ?
Would it make more sense?

Comment (alongosz) [inline eZ/Publish/Core/Repository/Values/ObjectState/ObjectState.php]: > What about instead deprecating property 'defaultLanguageCode', making it available via magic __get() for BC, and switch over to using 'mainLanguageCode' even for obj states and stategroups ?

Done in: 767ac9c followed by REST-related updates in b6a6f84 and 41538a2.
I kept `defaultLanguageCode` on [Create and Update Structs](https://github.com/ezsystems/ezpublish-kernel/tree/master/eZ/Publish/API/Repository/Values/ObjectState) as this seems to me like out of scope, but if you wish I could do that as well.
// cc @andrerom

File: eZ/Publish/API/Repository/Values/ObjectState/ObjectState.php [hunk 9]
```diff
@@ -18,7 +18,8 @@
  * @property-read mixed $id the id of the content type group
  * @property-read string $identifier the identifier of the content type group
  * @property-read int $priority the priority in the group ordering
- * @property-read string $defaultLanguageCode the default language of the object state group names and description used for fallback.
+ * @property-read string $mainLanguageCode the default language of the object state names and descriptions used for fallback.
+ * @property-read string $defaultLanguageCode deprecated, use $mainLanguageCode
```
Comment (andrerom) [inline eZ/Publish/API/Repository/Values/ObjectState/ObjectState.php]: `deprecated` => `@deprecated `

**Comment by alongosz** [comment]
Thanks for the merge :)

**Comment by andrerom** [comment]
Thanks for the PR's ;)
